src/main/scala/uk/gov/nationalarchives/MetadataService.scala

```diff
@@ -0,0 +1,100 @@
+package uk.gov.nationalarchives
+
+import cats.effect.IO
+import fs2.data.csv._
+import fs2.interop.reactivestreams._
+import fs2.{Chunk, Pipe, Stream, text}
+import uk.gov.nationalarchives.Lambda.Input
+import uk.gov.nationalarchives.MetadataService._
+
+import java.util.UUID
+
+class MetadataService(s3: DAS3Client[IO]) {
+  def parseCsvWithHeaders[T](input: Input, name: String)(implicit decoder: CsvRowDecoder[T, String]): IO[List[T]] = {
+    parseCsv(input, name, decodeUsingHeaders[T]())
+  }
+
+  def parseCsvWithoutHeaders[T](input: Input, name: String)(implicit decoder: RowDecoder[T]): IO[List[T]] = {
+    parseCsv(input, name, decodeWithoutHeaders[T](' '))
+  }
+
+  private def parseCsv[T](input: Input, name: String, decoderPipe: Pipe[IO, String, T]): IO[List[T]] = {
+    for {
+      pub <- s3.download(input.s3Bucket, s"${input.s3Prefix}$name")
+      csvString <- pub
+        .toStreamBuffered[IO](1024 * 5)
+        .flatMap(bf => Stream.chunk(Chunk.byteBuffer(bf)))
+        .through(text.utf8.decode)
+        .through(decoderPipe)
+        .compile
+        .toList
+    } yield csvString
+  }
+
+  def metadataToDynamoTables(
+      batchId: String,
+      departmentAndSeries: DepartmentSeries,
+      metadata: List[Metadata],
+      bagitManifests: List[BagitManifest]
+  ): IO[List[DynamoTable]] = {
+    val fileIdToChecksum: Map[UUID, String] = bagitManifests.map(bm => UUID.fromString(bm.filePath.stripPrefix("data/")) -> bm.checksum).toMap
+    val pathPrefix = departmentAndSeries.series
+      .map(series => s"${departmentAndSeries.department.id}/${series.id}")
+      .getOrElse(s"${departmentAndSeries.department.id}")
+    IO {
+      departmentAndSeries.department ::
+        metadata.map {
+          case AssetMetadata(identifier, parentPath, title) =>
+            DynamoTable(batchId, identifier, s"$pathPrefix/${parentPath.stripPrefix("/")}", title, "Asset", "", "")
+          case FileMetadata(identifier, parentPath, name, fileSize, title) =>
+            DynamoTable(
+              batchId,
+              identifier,
+              s"$pathPrefix/${parentPath.stripPrefix("/")}",
+              name,
+              "File",
+              title,
+              "",
+              Option(fileSize),
+              fileIdToChecksum.get(identifier),
+              name.split("\\.").lastOption
+            )
+          case FolderMetadata(identifier, parentPath, name, title) =>
+            val path = if (parentPath.isEmpty) pathPrefix else s"$pathPrefix/${parentPath.stripPrefix("/")}"
+            DynamoTable(batchId, identifier, path, name, "Folder", title, "")
+        } ++ departmentAndSeries.series.toList
+
+    }
+  }
+}
+object MetadataService {
+  trait Metadata
+
+  case class DynamoTable(
+      batchId: String,
+      id: UUID,
+      parentPath: String,
+      name: String,
+      `type`: String,
```

Comment:
Maybe `type` could be a sealed trait?

Reply:
Would editing the `toString` method (of each Type), to return a string representation of the class name, help?

Reply:
Also, could these be case objects instead of case classes and then you don't have to use the brackets?

Reply:
I've changed them to case objects and I've added a `toString` method to the trait. We still need the custom formatter because Scanamo turns any Scala object into a dynamo object but it's cleaner with the case objects and `toString` now instead of calling `getClass.getSimpleName`